### ISSUE TITLE
Add `is_truthy` and `is_falsy`

### ISSUE THREAD

**Issue by ThomasdenH** [rustwasm/wasm-bindgen#1632]
### Motivation

When porting Javascript, you'll frequently encounter `!value` and `!!value` constructs. They will return whether a value is "truthy" or "falsy". While this is not ideal for clean code, for incremental conversion it would be very handy to allow this in Rust as well.

### Proposed Solution

Inplement `Neg<Output = bool>` for `JsValue` to return whether the value is falsy.

```rust
let js_value: JsValue;
let bool_value: bool = !js_value;
```

### Alternatives

Since the operation is somewhat dirty, it might not be desirable to have it as a primary operator. Instead, it could be additional methods like `is_truthy` and `is_falsy`. This would be slightly less obvious to discover though,

### Additional Context

https://developer.mozilla.org/en-US/docs/Glossary/Truthy

**Comment by Pauan** [comment]
I like the idea of an `is_truthy` operator.

I also think we should have an `is_nully` operator (which would do `foo == null`). Though I don't like the name, so we should bikeshed that.

**Comment by alexcrichton** [comment]
I agree with @Pauan about having a function rather than an implementation of the `Neg` trait, and it seems like a good idea to me!

**Comment by ThomasdenH** [comment]
How would I implement this? I'd be interested in tackling this issue.

**Comment by ThomasdenH** [comment]
I have implemented a pure rust version. I don't know if it would be better to delegate this to JS though? If not, I can make a PR.

```rust
use wasm_bindgen::prelude::*;
use wasm_bindgen_test::*;
use wasm_bindgen::JsCast;
use js_sys::Number;

pub trait IsTruthy {
    fn is_falsy(&self) -> bool;
    fn is_truthy(&self) -> bool {
        !self.is_falsy()
    }
}

impl IsTruthy for JsValue {
    fn is_falsy(&self) -> bool {
        self.as_bool() == Some(false) ||
        self.as_string().map(|s| s.is_empty()).unwrap_or(false) ||
        self.is_null() ||
        self.is_undefined() ||
        self.dyn_ref::<Number>()
            .map(|number| number == &0u32 || Number::is_nan(&number))
            .unwrap_or(false)
    }
}

#[wasm_bindgen_test]
fn test_is_truthy() {
    assert_eq!(JsValue::from(0).is_truthy(), false);
    assert_eq!(JsValue::from(0).is_falsy(), true);
    assert_eq!(JsValue::from("".to_string()).is_truthy(), false);
    assert_eq!(JsValue::from("".to_string()).is_falsy(), true);
    assert_eq!(JsValue::from(false).is_truthy(), false);
    assert_eq!(JsValue::from(false).is_falsy(), true);
    assert_eq!(JsValue::NULL.is_truthy(), false);
    assert_eq!(JsValue::NULL.is_falsy(), true);
    assert_eq!(JsValue::UNDEFINED.is_truthy(), false);
    assert_eq!(JsValue::UNDEFINED.is_falsy(), true);

    assert_eq!(JsValue::from(10).is_truthy(), true);
    assert_eq!(JsValue::from(10).is_falsy(), false);
    assert_eq!(JsValue::from("null".to_string()).is_truthy(), true);
    assert_eq!(JsValue::from("null".to_string()).is_falsy(), false);
    assert_eq!(JsValue::from(true).is_truthy(), true);
    assert_eq!(JsValue::from(true).is_falsy(), false);
}
```

**Comment by alexcrichton** [comment]
This seems like a reasonable thing to delegate to JS engines, if not for performance if to just use their own definition. For that you could probably add a new intrinsic to test whether it looks like true or whether a value looks like false, and then basically just implement it in JS.

**Comment by Pauan** [comment]
@alexcrichton Should it be an intrinsic, or just `inline_js`?

```rust
#[wasm_bindgen(inline_js = "export function is_falsy(x) { return !x; }")]
extern "C" {
    fn is_falsy(x: &JsValue) -> bool;
}

impl JsValue {
    #[inline]
    pub fn is_falsy(&self) -> bool {
        is_falsy(self)
    }

    #[inline]
    pub is_truthy(&self) -> bool {
        !self.is_falsy()
    }
}
```

**Comment by alexcrichton** [comment]
That would also work quite well! And should perhaps be how we start implementing other intrinsics...

That may not be easily possible today though due to `inline_js` not being supported on node.js targets. For core intrinsics we may need to get that working first :(

**Comment by Pauan** [comment]
> That may not be easily possible today though due to inline_js not being supported on node.js targets. For core intrinsics we may need to get that working first :(

Right, but maybe we should consider it for new intrinsics.

----

@ThomasdenH So you can either go for the approach I gave above (using `inline_js`), or you can add a new intrinsic in all of these files:

* `crates/cli-support/src/intrinsic.rs`
* `crates/cli-support/src/js/rust2js.rs`
* `src/lib.rs`

(Search for `__wbindgen_is_undefined` and `IsUndefined` to see how existing intrinsics are defined)

**Comment by ThomasdenH** [comment]
Can I add doc tests?

**Comment by ThomasdenH** [comment]
This issue was implemented in #1638.
